File: Problem105.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

/**
 * Time Complexity - O(n)
 * Space Complexity - O(maxDepth) - recursion
 */
class Solution {

    TreeNode first = null;
    TreeNode second = null;

    TreeNode prev = null;

    public void recoverTree(TreeNode root) {
        if(root==null)
            return;
        helper(root);
        int temp = first.val;
        first.val = second.val;
        second.val = temp;
    }


    private void helper(TreeNode root){

        if(root==null)
            return ;

        helper(root.left);

        // Logic
        if(prev != null){
            if(first == null && prev.val >= root.val){
                first = prev;
            }

            if(first != null && prev.val >= root.val){
                second = root;
            }

        }

        prev = root;

        helper(root.right);
    }
}

File: Problem_105.py
# TC : O(N)
# SC : O(N)

"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        self.levels = [[]]
        def dfs(root, level):
            if not root:
                return
            if level == len(self.levels):
                self.levels.append([])
            self.levels[level].append(root)

            dfs(root.left, level+1)
            dfs(root.right, level+1)

        dfs(root, 0)
        for i in range(0, len(self.levels)):
            for j in range(1, len(self.levels[i])):
                self.levels[i][j-1].next = self.levels[i][j]

        return root
